### PR TITLE
Added GCP disks support

### PR DESCRIPTION
**Pull request by saratkumar-yb**

### CODE REVIEW

**Comment by mbautin** [comment]
The PR needs to have a better title and description.

File: crc.py [hunk 1]
```diff
@@ -146,8 +147,13 @@ def _get_ip(
...
         )
 
+    def slack_lookup_user_by_email(self, email):
+
+        user_info = self.slack_client.users_lookupByEmail(email=email)
+        return user_info["user"]["id"]
+
     def get_msg(
-        self, resource: str, operation_type: str, operated_list: List[str]
+        self, resource: str, operation_type: str, operated_list: object
```
Comment (mbautin) [inline crc.py:156]: Use `Any` instead of `object`.
If it could be multiple different types, use `Union`, etc.

File: crc.py [hunk 2]
```diff
@@ -161,12 +167,28 @@ def get_msg(
...
+            for key in operated_list.keys():
+                operated_list_length += len(operated_list[key])
+            if self.dry_run:
+                msg = f"`Dry Run`: Will be {operation_type}: `{operated_list_length}` {self.cloud} {resource}(s):\n"
+            else:
+                msg = f"{operation_type} the following `{operated_list_length}` {self.cloud} {resource}(s):\n"
 
-        return f"{operation_type} the following `{operated_list_length}` {self.cloud} {resource}(s):\n`{operated_list}`"
+            for key in operated_list.keys():
+                #member_id = self.slack_lookup_user_by_email(f"{key}@yugabyte.com")
```
Comment (mbautin) [inline crc.py:189]: Remove commented-out code.

File: crc.py [hunk 3]
```diff
@@ -365,6 +387,10 @@ def delete_disks(
         filter_tags: Dict[str, List[str]],
         exception_tags: Dict[str, List[str]],
         age: Dict[str, int],
+        detach_age: Dict[str, int],
+        name_regex: List[str],
+        exception_regex: List[str],
+        slack_notify_users,
```
Comment (mbautin) [inline crc.py:393]: Specify type for this parameter.

File: crc.py [hunk 4]
```diff
@@ -374,13 +400,27 @@ def delete_disks(
         :param exception_tags: Dictionary of tags to exclude the disks.
         :param age: Dictionary of age conditions to filter the disks.
         """
-        if self.cloud != "azure":
+        if self.cloud not in ["azure", "gcp"]:
             raise ValueError(
-                "Incorrect Cloud Provided. Disks operation is supported only on AZURE. AWS, GCP clean the NICs, Disks along with VM"
+                "Incorrect Cloud Provided. Disks operation is supported only on AZURE, GCP. AWS, GCP clean the NICs, Disks along with VM"
```
Comment (mbautin) [inline crc.py:405]: This is confusing. GCP is now mentioned twice in two sentences that were previously referring to mutually exclusive situations.

**Review by mbautin — CHANGES_REQUESTED** [review]

**Review by mbautin — CHANGES_REQUESTED** [review]

File: crc.py [hunk 5]
```diff
@@ -530,6 +604,20 @@ def get_argparser():
...
+        "--slack_notify_users",
+        action="store_true",
+        help="If true notify users in the Slack channel, currently only for GCP disk",
+    )
+
+    # Add Argument for label that need to be used for getting username
+    parser.add_argument(
+        "--slack_user_label",
+        metavar="SLACK_USER_LABEL",
+        help="The the gcp label that can be used to get username. Example: --slack_user_label yb_owner",
```
Comment (zlareb1-yb) [inline crc.py:618]: replace yb_owner with something like owner. This may leak sensitive tag information which we use in opensource

File: crc.py [hunk 6]
```diff
@@ -146,8 +149,23 @@ def _get_ip(
             f"Invalid cloud provided: {self.cloud}. Supported clouds are {CLOUDS}"
         )
 
+    def slack_lookup_user_by_email(self, email):
```
Comment (zlareb1-yb) [inline crc.py:152]: add docstring

File: crc.py [hunk 7]
```diff
@@ -146,8 +149,23 @@ def _get_ip(
...
+        try: 
+            user_info = self.slack_client.users_lookupByEmail(email=email)
+            return user_info["user"]["id"]
+        except:
+            return "not_found"
+        
+    def get_user_groups_list(self):
```
Comment (zlareb1-yb) [inline crc.py:159]: add docstring

File: crc.py [hunk 8]
```diff
@@ -374,13 +425,28 @@ def delete_disks(
         :param exception_tags: Dictionary of tags to exclude the disks.
         :param age: Dictionary of age conditions to filter the disks.
```
Comment (zlareb1-yb) [inline crc.py:426]: add added param information

File: crc.py [hunk 9]
```diff
@@ -374,13 +425,28 @@ def delete_disks(
...
         """
-        if self.cloud != "azure":
+        if self.cloud not in ["azure", "gcp"]:
             raise ValueError(
-                "Incorrect Cloud Provided. Disks operation is supported only on AZURE. AWS, GCP clean the NICs, Disks along with VM"
+                "Incorrect Cloud Provided. Disks operation is supported only on AZURE, and GCP. AWS clean the NICs, Disks along with VM"
```
Comment (zlareb1-yb) [inline crc.py:430]: Correct statement --> Incorrect Cloud Provided. Disks operation is supported only on AZURE and GCP. AWS cleans the NICs, Disks along with VM

File: crc.py [hunk 10]
```diff
@@ -662,6 +753,11 @@ def main():
         raise ValueError(
             "All Resources cleanup is supported only with all Clouds. Format: --cloud all --resources all"
         )
+    
+    if slack_notify_users and slack_user_label == None:
```
Comment (zlareb1-yb) [inline crc.py:757]: change this to --> if slack_notify_users and not slack_user_label:

File: crc/gcp/disk.py [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+    def count(self):
+        """
+        This is a property decorator that returns the count of items in the disk_names_to_delete list.
+        It's a read-only property, which means it can be accessed like a variable, but cannot be set like a variable.
+        """
+        count = len(self.disk_names_to_delete)
+        logging.info(f"count of items in disk_names_to_delete: {count}")
+        return count
+
+    def _is_old_disk_detach(self, disk):
```
Comment (zlareb1-yb) [inline crc/gcp/disk.py:105]: change method name to --> _is_disk_older_than_detach_age

File: crc/gcp/disk.py [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+                logging.info(
+                    f"number of GCP disks deleted: {len(self.disk_names_to_delete)}"
+                )
+                logging.warning(
+                    f"List of GCP disk deleted: {self.disk_names_to_delete}"
+                )
+            else:
+                logging.warning(
+                    f"List of GCP disk (Total: {len(self.disk_names_to_delete)}) which will be deleted: {self.disk_names_to_delete}"
+                )
```
Comment (zlareb1-yb) [inline crc/gcp/disk.py:207]: add new line

File: crc/gcp/disk.py [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+                        except Exception as e:
+                            # log the error message if an exception occurs
+                            logging.error(
+                                f"Error occurred while deleting disk {disk.name}: {e}"
+                            )
+            if not self.disk_names_to_delete:
+                logging.warning(f"No GCP disk to delete.")
+
+            if not self.dry_run:
+                logging.info(
```
Comment (zlareb1-yb) [inline crc/gcp/disk.py:198]: both can be logging.warning

File: crc/utils.py [hunk 14]
```diff
@@ -23,6 +24,9 @@ def init_logging(filename: str, log_level: str = "debug"):
     # Create a log message formatter.
     formatter = logging.Formatter(LOG_FORMATTER)
 
+    # Create log directory in script path if not found
+    os.makedirs(os.path.dirname(os.path.abspath(filename)), exist_ok=True)
+
     # Create file handler which logs debug messages to a log file.
     fh = logging.FileHandler(filename, mode="w")
     fh.setLevel(logging.INFO)
```
Comment (zlareb1-yb) [inline crc/utils.py:32]: checkout the crc.log file

**Review by zlareb1-yb — CHANGES_REQUESTED** [review]

File: crc/gcp/disk.py [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+    The default_disk_state variable specifies the default state of disks when querying for disks.
+    """
+
+    def __init__(
+        self,
+        dry_run: bool,
+        project_id: str,
+        filter_labels: Dict[str, List[str]],
+        exception_labels: Dict[str, List[str]],
+        age: int,
```
[the remark this answers is not in the document]
Comment (saratkumar-yb) [reply crc/gcp/disk.py:47]: will take rename as an improvement

File: crc/gcp/disk.py [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+        :type project_id: str
+        :param filter_labels: Dictionary of labels and their values used to filter Disks for deletion.
+        :type filter_labels: Dict[str, List[str]]
+        :param exception_labels: Dictionary of labels and their values used to exclude Disks from deletion.
+        :type exception_labels: Dict[str, List[str]]
+        :param age: Age in days of Disks that will be deleted.
+        :type age: int
+        :param detach_age: Age in days the Disks last got detached .
+        :type detach_age: int
+        :param notags: dictionary containing key-value pairs as filter tags to exclude disks which do not have these tags
```
[the remark this answers is not in the document]
Comment (saratkumar-yb) [reply crc/gcp/disk.py:66]: will take rename as a seprate improvement

File: crc/gcp/disk.py [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+        self,
+        dry_run: bool,
+        project_id: str,
+        filter_labels: Dict[str, List[str]],
+        exception_labels: Dict[str, List[str]],
+        age: int,
+        detach_age: int,
+        notags: Dict[str, List[str]],
+        name_regex: List[str],
+        exception_regex: List[str],
```
[the remark this answers is not in the document]
Comment (saratkumar-yb) [reply crc/gcp/disk.py:51]: will take rename as a seprate improvement

File: crc/gcp/disk.py [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+    The default_disk_state variable specifies the default state of disks when querying for disks.
+    """
+
+    def __init__(
+        self,
+        dry_run: bool,
+        project_id: str,
+        filter_labels: Dict[str, List[str]],
+        exception_labels: Dict[str, List[str]],
+        age: int,
```
Comment (zlareb1-yb) [inline crc/gcp/disk.py:47]: what about Azure when these params are not passed? like detach_age

**Comment by zlareb1-yb** [comment]
Add GitHub README changes as well

File: crc/gcp/disk.py (hunk 18 above)
Comment (saratkumar-yb) [reply crc/gcp/disk.py:47]: This should not effect Azure as detach_age param is optional. Will update README that currently we support only for GCP disks.

File: crc/utils.py (hunk 14 above)
Comment (saratkumar-yb) [reply crc/utils.py:32]: I have removed crc.log file from repo and creating the directory and file during run.

File: crc/gcp/disk.py (hunk 18 above)
Comment (saratkumar-yb) [reply crc/gcp/disk.py:47]: Tried a sample dry run of azure, works fine.

**Review by zlareb1-yb — APPROVED** [review]

File: README.md [hunk 19]
```diff
@@ -148,7 +152,7 @@ python crc.py --cloud azure --resource vm --filter_tags "{'test_task': ['stress-
 
 3. To delete all GCP disks that are older than 2 days and have the tag `test_task` with the value `stress-test` and project_id as 'test_project':
```
Comment (zlareb1-yb) [inline README.md:153]: make this 14 days instead of 2 days

File: crc.py [hunk 20]
```diff
@@ -373,14 +436,34 @@ def delete_disks(
...
             raise ValueError(
-                "Incorrect Cloud Provided. Disks operation is supported only on AZURE. AWS, GCP clean the NICs, Disks along with VM"
+                "Incorrect Cloud Provided. Disks operation is supported only on AZURE and GCP. AWS cleans the NICs, Disks along with VM"
             )
-
-        disk = Disk(self.dry_run, filter_tags, exception_tags, age, self.notags)
-        disk.delete()
+        if self.cloud == "azure":
+            disk = Disk(self.dry_run, filter_tags, exception_tags, age, self.notags)
+            disk.delete()
```
Comment (zlareb1-yb) [inline crc.py:451]: disk.delete() can be outside if else as it is common to both AZU, GCP

File: README.md (hunk 19 above)
Comment (saratkumar-yb) [reply README.md]: done

File: crc.py (hunk 20 above)
Comment (saratkumar-yb) [reply crc.py]: done

**Pull request merged**